Source/Dafny/DafnyAst.cs

```diff
@@ -325,30 +325,52 @@ public class BuiltIns
       return new ArrowType(f.tok, atd, f.Formals.ConvertAll(arg => Resolver.SubstType(arg.Type, typeMap)), Resolver.SubstType(f.ResultType, typeMap));
     }
 
-    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType) {
+    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType, List<bool> argumentGhostness = null) {
       Contract.Requires(tok != null);
       Contract.Requires(0 <= dims);
+      Contract.Requires(argumentGhostness == null || argumentGhostness.Count == dims);
       Contract.Ensures(Contract.Result<TupleTypeDecl>() != null);
 
       TupleTypeDecl tt;
-      if (!tupleTypeDecls.TryGetValue(dims, out tt)) {
+      argumentGhostness = argumentGhostness ?? new bool[dims].Select(_ => false).ToList();
+      if (!tupleTypeDecls.TryGetValue(argumentGhostness, out tt)) {
         Contract.Assume(allowCreationOfNewType);  // the parser should ensure that all needed tuple types exist by the time of resolution
-        if (dims == 2) {
-          // tuple#2 is already defined in DafnyRuntime.cs
-          tt = new TupleTypeDecl(dims, SystemModule, DontCompile());
-        } else {
-          tt = new TupleTypeDecl(dims, SystemModule, null);
-        }
-        tupleTypeDecls.Add(dims, tt);
+        // tuple#2 is already defined in DafnyRuntime.cs
+        var attributes = dims == 2 && !argumentGhostness.Contains(true) ? DontCompile() : null;
```

Comment:
Just checking, shouldn't this condition **be** something like 

`argumentGhostness.Where(x => !x).Sum() == 2` ?

Reply:
Good question. In this PR a Dafny tuple of type `(int, int, ghost int)` is compiled in C# to a `Tuple3OOG<T1, T2>` class, so it's not the same as compiling `(int, int)` and the `!argumentGhostness.Contains(true)` is correct. However, I just noticed that`(int, int, ghost int)` is compiled in Java to a `Tuple2<T1, T2>` class, because that compiler ignores the `TupleTypeDecl::CompileName` implementation (arguably, a bug of the compiler). Moreover, after [this decision](https://github.com/dafny-lang/dafny/pull/1269#issuecomment-877360472) in C# classes like`Tuple3OOG<T1, T2>` and `Tuple3OGO<T1, T2>` have exactly the same implementation and they only differ for the name.

@RustanLeino I remember we decided in a meeting to translate `(int, int, ghost int)` to a class `Tuple3OOG<T1, T2>`, but maybe we should reconsider and translate to `Tuple2<T1, T2>`. If we want the latter, it might be the first case in which different top level declarations gets compiled to the same class in the target language, so I might have to add a check to avoid declaring the same class multiple times.

Reply:
Let's leave it as `Tuple3OOG` and `Tuple3OGO` etc. in this PR. I suggest we change these to use `Tuple2` in the same future PR where we make the optimization that `Tuple1G<T>` would turn into just `T`.